### ISSUE TITLE
Linux process path foolery: need path parsing and on_disk invalidation

### ISSUE THREAD

**Issue by theopolis** [facebook/osquery#1607]
The Linux `processes` table includes the [`readlink`](http://linux.die.net/man/2/readlink) string response, which will include the " (deleted)" portion if the inode does not exist. This is interesting and potentially non-intended as if the filesystem path is changed (via a `mv`) the proc entry will begin reporting the new path (as long as the inode is consistent).

Furthermore, if the inode is unlinked and a new path is created at "/old/path (deleted)" the osquery `on_disk` column reports TRUE, when it should be FALSE.

**Comment by theopolis** [comment]
This is pretty hilarious:
```
$ cp `which osqueryi` /tmp/osqueryi
osquery> select pid, path, on_disk from processes where path like '%osqueryi%';
+---------+---------------+---------+
| pid     | path          | on_disk |
+---------+---------------+---------+
| 1860880 | /tmp/osqueryi | 1       |
+---------+---------------+---------+
[...] meanwhile...
$ rm /tmp/osqueryi
osquery> select pid, path, on_disk from processes where path like '%osqueryi%';
+---------+-------------------------+---------+
| pid     | path                    | on_disk |
+---------+-------------------------+---------+
| 1860880 | /tmp/osqueryi (deleted) | 0       |
+---------+-------------------------+---------+
[...] meanwhile...
$ cp `which osqueryi` /tmp/osqueryi\ \(deleted\)  
osquery> select pid, path, on_disk from processes where path like '%osqueryi%';
+---------+-------------------------+---------+
| pid     | path                    | on_disk |
+---------+-------------------------+---------+
| 1860880 | /tmp/osqueryi (deleted) | 1       |
+---------+-------------------------+---------+
```

**Comment by theopolis** [comment]
I'm not sure if this is an osquery problem or something more sinister?
```
/proc/1860880  » md5sum exe                                                                                                                                                                         reed@dev7360
f06a372730a442564c1994780d0cf857  exe
------------------------------------------------------------
/proc/1860880  » md5sum /tmp/osqueryi\ \(deleted\)                                                                                                                                                  reed@dev7360
f06a372730a442564c1994780d0cf857  /tmp/osqueryi (deleted)
```

**Comment by theopolis** [comment]
To "harden" the `on_disk` check, if a readlink path in `/proc/%d` contains a "(deleted)", we can choose to check the inode of the symlink target against the list in the processes memory map (also from proc).

**Comment by theopolis** [comment]
Not too worried about this at the moment.

**Comment by theopolis** [comment]
This needs some attention, at the very least "(deleted)" needs to be removed from the path string.

**Comment by zwass** [comment]
I'm going to dig into this.

I noticed while playing with the `processes` table that `on_disk` is `null` when it should be 0:

```
osquery> select pid, path, on_disk from processes where path like "%osquery%";
+------+---------------+---------+
| pid  | path          | on_disk |
+------+---------------+---------+
| 7555 | /tmp/osqueryi | 1       |
+------+---------------+---------+
osquery> ^Z
[1]+  Stopped                 /tmp/osqueryi
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ rm /tmp/osqueryi
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ fg
/tmp/osqueryi
select pid, path, on_disk from processes where path like "%osquery%";
+------+-------------------------+---------+
| pid  | path                    | on_disk |
+------+-------------------------+---------+
| 7555 | /tmp/osqueryi (deleted) |         |
+------+-------------------------+---------+
```

This looks like a bug, and I will fix it while I'm in here.

To address the problem when the same executable name ending with ` (deleted)` is in the directory, what do you think about comparing the second field in `/proc/%d/stat` (according to the man page contains "The filename of the executable, in parentheses")? This field does not seem to change like the symlinks do after the deletion.

**Comment by theopolis** [comment]
Thanks for looking into the `NULL`ish problem!

Comparing the process's `stat` details to the symlink-like textual pointer to where the filesystem thinks the INODE is and the weird `(deleted)` tag is a great idea! What about the case when the INODE link is moved (try moving your `/tmp/osqueryi` to `/tmp/osqueryi2`)? Would you expect `on_disk` to return `0`?

**Comment by zwass** [comment]
Actually, looks like the OS updates `/proc/%d/exe` appropriately. We get the new correct path and `on_disk` is still `1`.

(This is with the `NULL` fix, but no other new changes)

```
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ cp /bin/cat /tmp/cat
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ /tmp/cat
^Z
[2]+  Stopped                 /tmp/cat
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ ./build/trusty/osquery/osqueryi
osquery - being built, with love, at Facebook
Using a virtual database. Need help, type '.help'
osquery> select pid, path, name, on_disk from processes where name like "%cat%";
+------+----------+------+---------+
| pid  | path     | name | on_disk |
+------+----------+------+---------+
| 9720 | /tmp/cat | cat  | 1       |
+------+----------+------+---------+
osquery> ^Z
[3]+  Stopped                 ./build/trusty/osquery/osqueryi
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ mv /tmp/cat /tmp/cat2
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ fg
./build/trusty/osquery/osqueryi
select pid, path, name, on_disk from processes where name like "%cat%";
+------+-----------+------+---------+
| pid  | path      | name | on_disk |
+------+-----------+------+---------+
| 9720 | /tmp/cat2 | cat  | 1       |
+------+-----------+------+---------+
osquery> ^Z
[3]+  Stopped                 ./build/trusty/osquery/osqueryi
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ rm /tmp/cat2
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ fg
./build/trusty/osquery/osqueryi
select pid, path, name, on_disk from processes where name like "%cat%";
+------+---------------------+------+---------+
| pid  | path                | name | on_disk |
+------+---------------------+------+---------+
| 9720 | /tmp/cat2 (deleted) | cat  | 0       |
+------+---------------------+------+---------+
```

**Comment by zwass** [comment]
Ah, but `/proc/%d/stat` is not updated as we would hope (it still says `cat` and not `cat2`):

```
vagrant@vagrant-ubuntu-trusty-64:~$ cat /proc/9720/stat
9720 (cat) T 2825 9720 2825 34816 9721 1077960704 227 0 0 0 0 0 0 0 20 0 1 0 4796163 6074368 153 18446744073709551615 4194304 4238788 140731286885184 140731286884536 140586639566864 0 0 0 0 18446744071579340635 0 0 17 0 0 0 0 0 0 6336016 6337300 38113280 140731286886612 140731286886621 140731286886621 140731286888431 0
```

**Comment by zwass** [comment]
It doesn't look like my above proposal will work to detect the deletion. In this scenario with a symlink, the `name` in `/proc/%d/stat` comes from the name of the link (`dog`), while the path follows the link through (`/bin/cat`). If `/bin/cat` were deleted, comparing with `dog` would not be helpful.

```
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ ln -s /bin/cat /tmp/dog
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ /tmp/dog &
[2] 9744
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ ./build/trusty/osquery/osqueryi
osquery - being built, with love, at Facebook
Using a virtual database. Need help, type '.help'
osquery> select pid, path, name, on_disk from processes where name like "%cat%";
osquery> select pid, path, name, on_disk from processes where name like "%dog%";
+------+----------+------------+---------+
| pid  | path     | name       | on_disk |
+------+----------+------------+---------+
| 12   |          | watchdog/0 | -1      |
| 9744 | /bin/cat | dog        | 1       |
+------+----------+------------+---------+
osquery> ^Z
[2]-  Stopped                 /tmp/dog

[3]+  Stopped                 ./build/trusty/osquery/osqueryi
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ cat /proc/9744/stat
9744 (dog) T 2825 9744 2825 34816 9750 1077960704 276 0 0 0 0 0 0 0 20 0 1 0 4846885 6074368 153 18446744073709551615 4194304 4238788 140727541050832 140727541050184 139928517691408 0 0 0 0 18446744071579340635 0 0 17 0 0 0 0 0 0 6336016 6337300 21647360 140727541057748 140727541057757 140727541057757 140727541059567 0
```

**Comment by theopolis** [comment]
Instead of testing with soft links, try to `mv` the binary while a respective process is running.

The program will still be memory mapped. I am not sure if the INODE continues to be reserved/allocated or if it immediately returns to the free pool. I believe it's immediately marked free and the only persistence is the process(es)'s memory.

In this case, even though the binary is technically on a filesystem (disk?) the `on_disk` boolean seems like it should be `0`. If there's a way to detect the move, or alternatively a delete, via string comparisons of `./exe` and `./stat`, that would be helpful.

**Comment by zwass** [comment]
I did test `mv` separately from softlinks, see https://github.com/facebook/osquery/issues/1607#issuecomment-231166044.

In addition, it seems that the `/proc/%d/maps` file is updated appropriately as well:

```
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ touch /tmp/cat2\ \(deleted\)
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ cp /bin/cat /tmp/cat
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ /tmp/cat &
[1] 10230
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ head -1 /proc/10230/maps
00400000-0040b000 r-xp 00000000 08:01 64165                              /tmp/cat

[1]+  Stopped                 /tmp/cat
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ mv /tmp/cat /tmp/cat2
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ head -1 /proc/10230/maps
00400000-0040b000 r-xp 00000000 08:01 64165                              /tmp/cat2
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ rm /tmp/cat2
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ head -1 /proc/10230/maps
00400000-0040b000 r-xp 00000000 08:01 64165                              /tmp/cat2 (deleted)
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ ls -li /tmp/cat2\ \(deleted\)
58246 -rw-rw-r-- 1 vagrant vagrant 0 Jul  7 19:48 /tmp/cat2 (deleted)
```

Note at the end here that the inode appearing in the `maps` file and the inode for the spoof file are different. This suggests that the approach you described upthread comparing these inodes should work.

Is it reasonable enough to assume that the first line of the `maps` file will always point to the executable?

**Comment by theopolis** [comment]
> I did test mv separately from softlinks, see #1607 (comment).

Ah! Sorry I missed that :/

So in that test you've induced part of the weirdness. When the `./exe` (what the table is parsing) points to `/tmp/cat2`, will the `./stat` have that same path?

**Comment by zwass** [comment]
`./stat` unfortunately doesn't contain any sort of full path, just the "name".

`./maps` does contain the full path (along with inode), and it updates to match `./exe`. See my example above in https://github.com/facebook/osquery/issues/1607#issuecomment-231188070.

AFAICT the OS correctly handles operations that move or delete the file. When the binary file is moved, `./exe` and `./maps` are updated with the new file path. When the binary file is deleted (or another file is `mv`ed over it), ` (deleted)` is appended to the file path.

IMO in the situation in which the file is moved, we want to report the new path, and `on_disk = 1`. In the situation in which the file is deleted, we should report the path minus ` (deleted)`, and `on_disk = 0`. The only thing we should need to do here is use the inode stored in `./maps` to check whether the user is actually executing a binary that ends with ` (deleted)`.

Does this answer your questions? And the original concern?

**Comment by theopolis** [comment]
Sounds perfect!

**Comment by zwass** [comment]
This is fixed with #2234.
